### ISSUE TITLE
Invalid Property Value class __NSArrayM

### ISSUE THREAD

**Issue by YMMBilal** [Appboy/appboy-segment-ios#7]
Hi,

When I want to send a `NSArray` on an event, this `NSArray` is converted to a `NSMutableArray` by `SEGCoerceJSONObject` (from `SEGAnalyticsUtils.m`).
Yet, Appboy does not handle `NSMutableArray` but `NSArray`. So the Xcode debugger prints the following information :
>[APPBOY] Invalid Property Value class __NSArrayM for key allergies removing.

Can you fix it ?

Thanks in advance,
Bilal

**Comment by chshapiro** [comment]
Hi @YMMBilal ,

Thank you for reaching out and reporting this issue. 

Are you trying to pass an NSArray with key "allergies" as a property on an event?

For custom event properties, we only support properties of types NSString, NSNumber, and NSDate. We don't support properties of any type of array.

Please reach out out if you have any further questions or if this is not the issue you were having.

Thank you,
Chava

**Comment by f2prateek** [comment]
hey @chshapiro, the segment spec does allow for arrays (e.g. https://segment.com/docs/spec/ecommerce/v2/#product-list-viewed).

If Appboy's SDK doesn't accept arrays, then it should filter those properties it doesn't handle at the integration level. If it does not accept NSMutableArray/NSMutableDictionary (but accepts NSArray/NSDictionary), then it should convert those in the integration to their immutable counterparts.

I don't have a an example for iOS on hand, but here's an example for analytics.js https://github.com/fullstorydev/analytics.js-integration-fullstory/pull/10. FullStory's javascript library does not allow for arrays/nested objects, so they skip those properties so that the event and other acceptable properties are still processed.

**Comment by chshapiro** [comment]
Hi @f2prateek,

Thank you for the advice!  Currently when an event is passed down with an array property, the SDK tracks the event and logs all other valid properties, and filters out the invalid array property. We log this error message to let the developer know about the abandoned property. Is this the same as your suggestion?

If you have any further question or suggestions, please let us know. 
Thank you!

**Comment by f2prateek** [comment]
@chshapiro not quite. The current behaviour is that the AppBoy SDK rejects the properties, but since the Segment SDK sends NSMutableArrays, those will be rejected. To handle this correctly with Segment, the integration (this repo, not the AppBoy SDK) should convert NSMutableArray to NSArray.

**Comment by chshapiro** [comment]
Hi @f2prateek 

Thanks for that info.  Currently our SDK does not allow any kind of array for properties, so we're not rejecting anything that could be transformed into valid data.

In general we definitely agree that we should transform data where possible to conform to our APIs.  However, we're wary of copying validation logic from our core SDK to this wrapper, as we suspect the value add isn't worth the maintenance burden.  Is there a compelling reason to port the validation logic here as well?

Thanks!

**Comment by f2prateek** [comment]
@chshapiro I see, could you clarify what happens if a customer calls `[analytics track:@"Some Event" properties:@{ @"some_valid_property": @"foo", @"some_array_property": @[ @"foo", @"bar" ] }];`.

My understanding is that because of `some_array_property`, AppBoy drops the `Some Event` entirely, instead of rejecting just the array property and accepting the event with `some_valid_property`.

If Appboy does reject the event entirely, I think the integration should be updated to reject just invalid properties but still accept valid ones. If AppBoy already accepts the partial event correctly there is nothing to be done.

I was originally a bit confused by the behaviour @YMMBilal described:
> Appboy does not handle NSMutableArray but NSArray

Which is why I suggested `NSMutableArray`s be converted to `NSArray` internally. But I understand now the case that AppBoy does not accept both NSMutableArray and NSArray properties. So rejecting the `NSArray` properties would be a good solution.

**Comment by qclmnt** [comment]
I don't really understand why Segment accept NSArray and not the Appboy integration. It should be available too... Otherwise you use an other integration with Array and that one with string... It is a little bit worrying.

**Comment by chshapiro** [comment]
Hi @qclmnt,

Our Segment integration aims to map Segment's interface to our SDK's inteface in a way that is as complete as possible.  

Given that Segment is routing requests from one interface to several other, different interfaces (developed separately), there are bound to be places where something is present in one interface and not another. 

If there's functionality that can be transformed to match our interface we're very happy to do that, as long as it's useful for every integrator (i.e. we don't want to add data that isn't useful to some integrations).  Otherwise, folks can either call our SDK directly or they can fork and modify our integration.

Please let us know if you have any further comments/questions.

Thanks!

**Comment by f2prateek** [comment]
@qclmnt to build on what @chshapiro said, the issue here isn't that the integration doesn't accept NSArray properties, but that the Appboy SDK itself doesn't accept NSArray properties.

**Comment by qclmnt** [comment]
@f2prateek we are able to send Array by our server integration, that's why I didn't understood

**Comment by f2prateek** [comment]
Interesting, @chshapiro is it true that the appboy server api accepts arrays but the client side appboy sdk does not?

**Comment by briancaw** [comment]
Hi @qclmnt and @f2prateek 

There's a few different places to consider:
- Our [core SDKs](https://github.com/appboy/appboy-ios-sdk) only support arrays in custom user attributes.
- Our [Rest API](https://www.appboy.com/documentation/REST_API/#user-attributes-object-specification) only supports arrays for custom user attributes (other arrays in the API are just batching data).
- Our Segment wrapper does `not` currently support custom user attribute arrays and we have an open issue to add support.

In short, our core interface supports arrays as data types only for custom attributes.  Our Segment is missing that aspect of the interface.  Nowhere supports event property arrays.

Please let me know if that all makes sense and any followup questions/comments.

Thanks,
Brian

**Comment by briancaw** [comment]
Hi All, 

We've added custom attribute array support to our issue queue for our next release.  We're aiming to get that out in Q1.  The reported issue was expected behavior, so closing for now.  Please reopen if you have any further questions/comments.

Thanks,
Brian

**Comment by bricepollock** [comment]
@briancaw can you provide a reference PR or something we can use to track this work? 

I don't see any open issues or PRs that address this?

**Comment by Bucimis** [comment]
@bricepollock Custom attribute array support was added in this commit https://github.com/Appboy/appboy-segment-ios/commit/9625d925bcab2ed58f4660637e042efc9ccef0b5#r38491680 (v2.0.0 https://github.com/Appboy/appboy-segment-ios/blob/master/CHANGELOG.md#200)

**Comment by bricepollock** [comment]
Hello @Bucimis and @briancaw it looks like this PR was pulled into 3.0.0.

So I updated my pods
```
       pod 'Appboy-iOS-SDK', '~> 3.26.1'
       pod 'Segment-Appboy', '~> 3.3.0'
```

However, I still see the errors after updating:
2020-08-25 12:45:44.151716-0700 betterup[73506:7261755] [APPBOY] Invalid Property Value class Swift.__SwiftDeferredNSArray for key client_features_enabled. Not retaining.

Should I consider this a new issue?

**Comment by chshapiro** [comment]
Hey @bricepollock, 

This looks like an error you may be getting when trying to include an array in the `properties` dictionary for a custom event. That is not currently supported by the SDK. Is that something you are trying to do in your app?

**Comment by bricepollock** [comment]
Yes @chshapiro that is exactly the use case. I had thought the AppBoy SDK would handle all the properties that Segment can handle, but it seems not to be the case with arrays. 

Is there a way to disable these messages in console?

**Comment by chshapiro** [comment]
Hey @bricepollock,

Unfortunately the only way to disable this particular log would be to completely turn off logging which we do not recommend doing.

**Comment by bricepollock** [comment]
@chshapiro do you have it on your roadmap to support arrays or to provide an opt out of verbose logging? It fills the console to make it unreadable and Braze is the only tool that doesn't support these data.

**Comment by chshapiro** [comment]
Hey @bricepollock,

If you have verbose logging on, you can turn it on by following the [docs](https://www.braze.com/docs/developer_guide/platform_integration_guides/ios/initial_sdk_setup/other_sdk_customizations/#braze-log-level) but this particular log will be logged regardless unless you completely turn off all logging which is not recommended. 

The `Appboy-iOS-SDK` currently does not support custom event properties in the form of arrays. We don't have a timeline on supporting that but are tracking it internally. Once it's supported on the `Appboy-iOS-SDK`, we will then be able to support it on `Segment-Appboy`.

**Comment by bricepollock** [comment]
Can you provide a way to disable this log in the meantime then? Its not possible to pass braze separate properties since braze is integrated into our segment events.
